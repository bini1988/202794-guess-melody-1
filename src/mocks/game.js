
export const maxTime = 5;
export const maxMistakes = 3;
export const questions = [
  {
    type: `genre`,
    genre: `rock`,
    answers: [
      {
        src: `https://upload.wikimedia.org/wikipedia/commons/1/1f/Uganda_flag_and_national_anthem_-_Oh_Uganda_Land_o.ogg`,
        genre: `rock`,
      }, {
        src: `https://upload.wikimedia.org/wikipedia/commons/1/1f/Uganda_flag_and_national_anthem_-_Oh_Uganda_Land_o.ogg`,
        genre: `blues`,
      }, {
        src: `https://upload.wikimedia.org/wikipedia/commons/1/1f/Uganda_flag_and_national_anthem_-_Oh_Uganda_Land_o.ogg`,
        genre: `jazz`,
      }, {
        src: `https://upload.wikimedia.org/wikipedia/commons/1/1f/Uganda_flag_and_national_anthem_-_Oh_Uganda_Land_o.ogg`,
        genre: `rock`,
      },
    ],
  }, {
    type: `artist`,
    song: {
      artist: `Jim Beam`,
      src: `https://upload.wikimedia.org/wikipedia/commons/1/1f/Uganda_flag_and_national_anthem_-_Oh_Uganda_Land_o.ogg`,
    },
    answers: [
      {
        picture: `https://i.pravatar.cc/134x134`,
        artist: `John Snow`,
      }, {
        picture: `https://i.pravatar.cc/134x134`,
        artist: `Jack Daniels`,
      }, {
        picture: `https://i.pravatar.cc/134x134`,
        artist: `Jim Beam`,
      },
    ],
  },
];
