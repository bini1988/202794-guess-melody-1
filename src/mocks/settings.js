
export const maxTime = 5;
export const maxMistakes = 3;
