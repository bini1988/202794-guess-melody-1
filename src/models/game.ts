/* eslint-disable no-invalid-this */
import {observable, action, flow} from "mobx";
import axios from "../axios";
import {ArtistAnswer, GenreAnswer, QuestionTypes, GameQuestion} from "../types.d";
import {questions} from "../mocks/game";

class Game {
  @observable public fetching: boolean = false;
  @observable public error = null;

  @observable public maxTime: number = 5;
  @observable public maxMistakes: number = 3;
  @observable public questions: GameQuestion[] = questions as GameQuestion[];
  @observable public questionIndex: number = 0;
  @observable public answers = [];
  @observable public mistakes: number = 5;
  @observable public time: number = 0;

  @action.bound
  public restart() {
    this.questionIndex = -1;
    this.mistakes = 0;
    this.time = 0;
  }

  @action.bound
  public handleStep() {
    if (this.mistakes >= this.maxMistakes) {
      this.restart();
    }
  }

  @action.bound
  public handleArtistAnswer(answer: ArtistAnswer, time: number) {
    const question = this.questions[this.questionIndex];

    if (question.type === QuestionTypes.Artist) {
      const isAnswerCorrect =
        (question.song.artist === answer.artist);

      this.handleAnswer(isAnswerCorrect, time);
    }
  }

  @action.bound
  public handleGenreAnswer(answers: GenreAnswer[], time: number) {
    const question = this.questions[this.questionIndex];

    if (question.type === QuestionTypes.Genre) {
      const byGenre = (it: GenreAnswer) => it.genre === question.genre;
      const answersCount = question.answers.filter(byGenre).length;
      const isGenresCorrect = answers.every(byGenre);
      const isAnswersCorrect =
        isGenresCorrect && (answers.length === answersCount);

      this.handleAnswer(isAnswersCorrect, time);
    }
  }

  @action.bound
  public fetchQuestions = flow(function* (this: Game) {
    this.fetching = true;
    this.error = null;

    try {
      const responce = yield axios.get(`/questions`);
      this.questions = responce.data;
    } catch (error) {
      this.error = error;
    } finally {
      this.fetching = false;
    }
  })

  @action
  private handleAnswer(isCorrect: boolean, time: number) {
    this.mistakes += +!isCorrect;
    this.time += time;
  }
}

export {Game};
export default new Game();
