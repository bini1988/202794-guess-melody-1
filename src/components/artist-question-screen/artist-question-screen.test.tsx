import React from "react";
import renderer from "react-test-renderer";
import serializer from "jest-emotion";
import {ArtistQuestion} from "../../types.d";
import {artistQuestion} from "../../mocks/game";
import ArtistQuestionScreen from "./artist-question-screen";

expect.addSnapshotSerializer(serializer);

it(`ArtistQuestionScreen correctly renders default markup`, () => {
  const tree = renderer
    .create(
        <ArtistQuestionScreen
          question={artistQuestion as ArtistQuestion}/>
    ).toJSON();

  expect(tree).toMatchSnapshot();
});
