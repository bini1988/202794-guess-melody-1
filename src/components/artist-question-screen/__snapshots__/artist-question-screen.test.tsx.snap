// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ArtistQuestionScreen correctly renders default markup 1`] = `
.emotion-32 {
  position: relative;
  width: 440px;
  margin: 0 auto;
  margin-top: 150px;
  margin-bottom: 105px;
}

.emotion-0 {
  margin: 0;
  margin-bottom: 30px;
  font-size: 33px;
  font-weight: 300;
  font-style: italic;
  color: #230d1a;
  text-align: center;
}

.emotion-7 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: space-around;
  -webkit-justify-content: space-around;
  -ms-flex-pack: space-around;
  justify-content: space-around;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.emotion-3 {
  padding: 0;
  width: 35px;
  height: 50px;
  border: none;
  background-color: transparent;
  -webkit-transition: -webkit-transform ease-in-out 300ms;
  -webkit-transition: transform ease-in-out 300ms;
  transition: transform ease-in-out 300ms;
  outline: none;
  cursor: pointer;
}

.emotion-3:not(:disabled):hover,
.emotion-3:not(:disabled):focus {
  -webkit-transform: scale(1.05);
  -ms-transform: scale(1.05);
  transform: scale(1.05);
}

.emotion-3:not(:disabled):active {
  -webkit-transform: scale(1.04);
  -ms-transform: scale(1.04);
  transform: scale(1.04);
}

.emotion-3 svg {
  fill: #ff9749;
  -webkit-filter: drop-shadow(3px 3px 3px rgba(0,0,0,0.2));
  filter: drop-shadow(3px 3px 3px rgba(0,0,0,0.2));
}

.emotion-2 {
  position: absolute;
  width: 1px;
  height: 1px;
  margin: -1px;
  border: 0;
  padding: 0;
  white-space: nowrap;
  -webkit-clip-path: inset(100%);
  clip-path: inset(100%);
  -webkit-clip: rect(0 0 0 0);
  clip: rect(0 0 0 0);
  overflow: hidden;
}

.emotion-5 {
  -webkit-flex: 0 1 auto;
  -ms-flex: 0 1 auto;
  flex: 0 1 auto;
  margin: 0 10px;
  height: 55px;
  width: 370px;
}

.emotion-5 > svg {
  -webkit-filter: drop-shadow(5px 5px 5px rgba(0,0,0,0.2));
  filter: drop-shadow(5px 5px 5px rgba(0,0,0,0.2));
}

.emotion-30 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  margin-top: 80px;
}

.emotion-14 {
  width: 134px;
  text-align: center;
  cursor: pointer;
}

.emotion-14:hover .emotion-11 {
  border-color: #ff9749;
}

.emotion-14:active .emotion-11 {
  border-color: #f00000;
}

.emotion-10 {
  box-sizing: border-box;
  width: 134px;
  height: 134px;
  margin-bottom: 5px;
  background-color: #f0eed5;
  border: solid 5px #230d1a;
  border-radius: 50%;
  overflow: hidden;
  cursor: pointer;
  box-shadow: 5px 5px 15px -5px #000;
  -webkit-transition: border-color ease-in-out 350ms;
  transition: border-color ease-in-out 350ms;
}

.emotion-10 img {
  max-width: 100%;
  height: auto;
  -webkit-transform: scale(1.1);
  -ms-transform: scale(1.1);
  transform: scale(1.1);
}

<section
  className="emotion-32 emotion-33"
>
  <h2
    className="emotion-0 emotion-1"
  >
    Кто исполняет эту песню?
  </h2>
  <div
    className="emotion-7 emotion-8"
  >
    <button
      className="emotion-3 emotion-4"
      disabled={false}
      onClick={[Function]}
      type="button"
    >
      <svg
        height="50"
        width="35"
      >
        <use
          xlinkHref="#play"
          xmlnsXlink="http://www.w3.org/1999/xlink"
        />
      </svg>
      <span
        className="emotion-2"
      >
        Проиграть трек
      </span>
    </button>
    <div
      className="emotion-5 emotion-6"
    >
      <svg
        height="55"
        width="370"
      >
        <use
          xlinkHref="#equalizer"
          xmlnsXlink="http://www.w3.org/1999/xlink"
        />
        <rect
          clipPath="url(#equalizer-lines)"
          fill="#ff9749"
          height="100%"
          width="0%"
        />
      </svg>
      <audio
        onCanPlayThrough={[Function]}
        onEnded={[Function]}
        onError={[Function]}
        onLoadStart={[Function]}
        onTimeUpdate={[Function]}
        src="https://upload.wikimedia.org/wikipedia/commons/d/db/Gimn_Sovetskogo_Soyuza_%281977_Vocal%29.oga"
      />
    </div>
  </div>
  <form
    className="emotion-30 emotion-31"
  >
    <div
      className="emotion-14 emotion-15"
    >
      <input
        className="emotion-2"
        id="artist-0"
        name="answer"
        onChange={[Function]}
        type="radio"
        value="John Snow"
      />
      <label
        className="emotion-12 emotion-13"
        htmlFor="artist-0"
      >
        <div
          className="emotion-10 emotion-11"
        >
          <img
            alt="John Snow"
            height="134"
            src="https://i.pravatar.cc/134x134"
            width="134"
          />
        </div>
        John Snow
      </label>
    </div>
    <div
      className="emotion-14 emotion-15"
    >
      <input
        className="emotion-2"
        id="artist-1"
        name="answer"
        onChange={[Function]}
        type="radio"
        value="Jack Daniels"
      />
      <label
        className="emotion-12 emotion-13"
        htmlFor="artist-1"
      >
        <div
          className="emotion-10 emotion-11"
        >
          <img
            alt="Jack Daniels"
            height="134"
            src="https://i.pravatar.cc/134x134"
            width="134"
          />
        </div>
        Jack Daniels
      </label>
    </div>
    <div
      className="emotion-14 emotion-15"
    >
      <input
        className="emotion-2"
        id="artist-2"
        name="answer"
        onChange={[Function]}
        type="radio"
        value="Jim Beam"
      />
      <label
        className="emotion-12 emotion-13"
        htmlFor="artist-2"
      >
        <div
          className="emotion-10 emotion-11"
        >
          <img
            alt="Jim Beam"
            height="134"
            src="https://i.pravatar.cc/134x134"
            width="134"
          />
        </div>
        Jim Beam
      </label>
    </div>
  </form>
</section>
`;
